### DISCUSSION TITLE
How to prepend env variable?

### DISCUSSION THREAD

**Discussion by wingyplus** [dagger/dagger#5001] [Help] [answered]
I want to install something and append it to `PATH` to make it accessible. So how to prepend the new directory into the env variable as we do with `PATH`?

**Comment by wingyplus** [comment]
I found one possible solution is query env variable PATH and use this value for appending with new value. I waiting for someone who give a better solution than I do

**Reply by kpenfound** [reply]
Hey @wingyplus , just to clarify, are you talking about updating the path in a container you're creating? Something like:
```go
client.Container()
...
.WithEnvVariable('PATH', '/root/.nvm/versions/node/v16.17.0/bin:$PATH')
```
where `$PATH` would evaluate (today the above does not work)

**Reply by wingyplus** [reply]
Yes, exactly.

**Reply by helderco** [reply, marked as answer]
Yes, you are correct:

```python
path = await ctr.get_variable("PATH")
print(await (ctr
  .with_env_variable("PATH", f"/opt/venv/bin:{path}")
  .with_exec(["sh", "-c", "echo $PATH"])
  .stdout()
))
```

**Reply by wingyplus** [reply]
Thank you, you all. 🙇

**Reply by wingyplus** [reply]
Mark @helderco as answer for clearer solution. 🙇

**Comment by shykes** [comment]
I think this might deserve its own step in the core API… It’s possible in userland, but pretty verbose, and the question does come up often. Seems worth the overhead of a builtin helper. Any thoughts on design?

**Reply by vito** [reply]
I agree; extending `$PATH` seems like a common use case. it would be nice to have parity with Dockerfile `ENV` which supports interpolation.

It might be a surprising default, since who knows what values people might try to set in env vars. An `expand: true` param would be simple enough to implement by using [`os.Expand`](https://pkg.go.dev/os#Expand) and fetching values from the current image config.

```graphql
container {
  withEnvVariable(name: "PATH", value: "/opt/venv/bin:${PATH}", expand: true) {
    withExec(args: []string{"sh", "-c", "echo $PATH"}) {
      stdout
    }
  }
}
```

(Interpolate might be a more obvious name.)

**Reply by helderco** [reply]
Yes, I second @vito's suggestion. 👍

**Reply by helderco** [reply]
As for doing it by default I'm not worried about the values people may be using, just that we now would have an exception to "we don't expand env vars". Some people try to use it in `WithExec()` without a shell too, like they can in a Dockerfile. Being explicit makes it  clearer. I expect users to ask for it to be included in other places as well when we add this.

I have an inkling of a memory for why we've never done it, from version v0.2, but don't remember the arguments.

**Reply by vito** [reply]
@helderco Here's my thinking re: default behavior:

Dockerfiles can get away with interpolating by default because the only way to pass in a dynamic value is through build args, which _also_ uses interpolation syntax. So either the interpolation is satisfied by an env var, or it is satisfied by a build arg:

```dockerfile
FROM golang
ENV PATH=/foo:$PATH
RUN echo $PATH
```

```dockerfile
FROM golang
ARG ARG=/bar
ENV PATH=/foo:$ARG
RUN echo $PATH
```

If you were to pass `--build-arg 'ARG=foo$bar'` the `$bar` wouldn't be expanded. Dagger doesn't have this luxury; you pass values straight in through the API, so it's impossible to know whether the `foo$bar` value you passed in was a parameter given to your Dagger script and passed along, or whether your Dagger script expected `$bar` to be expanded.

---

Side note: interestingly, env vars seem to take precedence.

```dockerfile
FROM golang
ARG PATH=/bar
ENV PATH=/foo:$PATH
RUN echo $PATH
```

```sh
$ docker build . --build-arg PATH=/baz --no-cache
# ...
Step 4/4 : RUN echo $PATH
 ---> Running in 24c6b84b69a4
/foo:/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

**Comment by marcosnils** [comment]
:wave: another user getting confused with this in Discord today: https://discord.com/channels/707636530424053791/1107670041362706532/1107670041362706532

Might want to bump the priority on this one. cc @d3rp3tt3 @shykes

**Reply by shykes** [reply]
Thanks for the bump. I escalated it to the community support backlog

**Comment by mircubed** [comment]
@wingyplus thank you for creating this post. I have transitioned the post into an issue to ensure that it gets on the community backlog. You can track it in #5140. 

Thanks everyone!

**Reply by wingyplus** [reply]
Thank you. 😊

**Reply by mircubed** [reply]
Note that the issue is now closed, please let us know if you run into any additional issues!
